File: WebApplication/ClientApp/src/app/_orchestration/getLeaderboard/getCourseLeaderboard/courseLeaderboardDto.ts
import { LocationInfoWithRank } from "src/app/_orchestration/locationInfoWithRank";
import { DisplayNameWithIdDto } from "../../displayNameWithIdDto";
import { BracketType } from "../../../_core/enums/bracketType";
import { IIntervalType } from "../../../_common/IIntervalName";
import { IRaceSeriesType } from "../../../_common/IRaceSeriesType";
import { LeaderboardResultDto } from "../leaderboardResultDto";

export interface CourseLeaderboardDto extends IRaceSeriesType {
  raceName: string
  raceSeriesDescription: string
  locationInfoWithRank: LocationInfoWithRank
  courseName: string
  courseDate: string
  courseTime: string
  courseDistance: number
  courseMetadata: CourseMetadata
  leaderboards: CourseLeaderboardByIntervalDto[]
}

export interface CourseLeaderboardByIntervalDto extends IIntervalType {
  intervalName: string
  totalRacers: number
  results: LeaderboardResultDto[]
}

export interface CourseMetadata {
  courses: DisplayNameWithIdDto[]
  brackets: BracketMetadata[]
  intervals: DisplayNameWithIdDto[]
}

export interface BracketMetadata extends DisplayNameWithIdDto {
  bracketType: BracketType
}

File: Api/Orchestration/GetLeaderboard/GetCourseLeaderboard/CourseLeaderboardDto.cs
using Api.DataModels.Enums;

namespace Api.Orchestration.GetLeaderboard.GetCourseLeaderboard;

public record CourseLeaderboardDto
{
    public required string CourseDate { get; init; }
    public required double CourseDistance { get; init; }
    public required string CourseName { get; init; }
    public required string CourseTime { get; init; }
    public required List<CourseLeaderboardByIntervalDto> Leaderboards { get; init; }
    public required LocationInfoWithRank LocationInfoWithRank { get; init; }
    public required int RaceId { get; init; }
    public required string RaceName { get; init; }
    public required int RaceSeriesId { get; init; }
    public required string RaceSeriesDescription { get; init; }
    public required RaceSeriesType RaceSeriesType { get; init; }
    public required string RaceSeriesTypeName { get; init; }
}
